File: lab5/zadanie2/python/zadanie2.py
from math import cos

def f(x):
	return cos((x)/2)

def findZero(f, a, b, eps):
	if f(a)*f(b) >= 0:
		print("W przedziale nie znajduje się miejsce zerowe!")
		return None
	if abs(b-a) < eps:
		return (a+b)/2
	while abs(b-a) >= eps:
		c = (a+b)/2
		if f(c) == 0:
			return c
		elif f(c) * f(a) < 0:
			b = c
		else:
			a = c
	return (b+a)/2

if __name__ == "__main__":
	a = float(input("Podaj a: "))
	b = float(input("Podaj b: "))
	eps = float(input("ε = 10^(-k). Podaj k: "))
	eps = 10**(-eps)

	result = findZero(f, a, b, eps)
	if result != None:
		print(result)

File: lab5/zadanie2/c/zadanie2.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <stdbool.h>

typedef double (*functype)(double);

double f(double x) {
	return cos(x/2);
}

double findZero(functype f, double a, double b, double eps){
	double c;
	if (fabs(b-a) < eps){
		return (a+b)/2;
	}
	while (fabs(b-a) >= eps){
		c = (a+b)/2;
		if (f(c) == 0){
			return c;
		}
		else if (f(c) * f(a) < 0){
			b = c;
		}
		else {
			a = c;
		}
	}
	return (a+b)/2;
}

bool hasZero(functype f, double a, double b){
	if (f(a) * f(b) >= 0){
		printf("W przedziale nie znajduje się miejsce zerowe!");
		return false;
	}
	return true;
}

int main(){
	double a, b, eps, result;
	printf("Podaj a: ");
	scanf("%lf", &a);
	printf("Podaj b: ");
	scanf("%lf", &b);
	printf("ε = 10^(-k). Podaj k: ");
	scanf("%lf", &eps);
	eps = pow(10, -eps);
	if (!hasZero(f, a, b)) {
		return -1;
	}
	result = findZero(f, a, b, eps);
	printf("%lf\n", result);
	return 0;
}
